trunk/SUAVE/Components/Energy/Networks/Propulsor_Surrogate.py

```diff
@@ -32,159 +34,322 @@ class Propulsor_Surrogate(Propulsor):
+            regr_sfc = linear_model.LinearRegression()          
+            sfc_surrogate  = regr_sfc.fit(xy, sfc)
+            thr_surrogate  = regr_thr.fit(xy, thr)
+            
+        else:
+            raise NotImplementedError('Selected surrogate method has not been implemented')
+       
+       
+        if self.thrust_anchor is not None:
+            cons = deepcopy(self.thrust_anchor_conditions)
+            cons[0,0] /= self.altitude_input_scale
+            base_thrust_at_anchor = thr_surrogate.predict(cons)
+            self.thrust_anchor_scale = self.thrust_anchor/(base_thrust_at_anchor*self.thrust_input_scale)
+            
+        if self.sfc_anchor is not None:
+            cons = deepcopy(self.sfc_anchor_conditions)
+            cons[0,0] /= self.altitude_input_scale
+            base_sfc_at_anchor = sfc_surrogate.predict(cons)
+            self.sfc_anchor_scale = self.sfc_anchor/(base_sfc_at_anchor*self.sfc_input_scale)
+       
         # Save the output
         self.sfc_surrogate    = sfc_surrogate
-        self.thrust_surrogate = thr_surrogate     
\ No newline at end of file
+        self.thrust_surrogate = thr_surrogate   
+        
+    def extended_thrust_surrogate(self, thr_surrogate, cond, lo_blender, hi_blender):
+        """ Fixes thrust values outside of the standard throttle range in order to provide
+            reasonable values outside of the typical surrogate coverage area. 
+            
+            Assumptions:
+            None
+            
+            Source:
+            N/A
+            
+            Inputs:
+            thr_surrogate     - Trained sklearn surrogate that outputs a scaled thrust value
+            cond              - nx3 numpy array with input conditions for the surrogate
+            lo_blender        - Cubic spline blending class that is used at the low throttle cutoff
+            hi_blender        - Cubic spline blending class that is used at the high throttle cutoff
+            
+            Outputs:
+            T                 [nondim]
+            
+            Properties Used:
+            None
+        """            
+        # initialize
+        cond_zero_eta = deepcopy(cond)
+        cond_one_eta = deepcopy(cond)
+        cond_zero_eta[:,2] = 0
+        cond_one_eta[:,2] = 1
+        min_thrs = thr_surrogate.predict(cond_zero_eta)
+        max_thrs = thr_surrogate.predict(cond_one_eta)
+        dTdetas = max_thrs - min_thrs
+        etas = cond[:,2]
+        mask_low = etas < 0
+        mask_lo_blend = np.logical_and(etas >= 0, etas < 0.01)
+        mask_mid = np.logical_and(etas >= 0.01, etas < 0.99)
+        mask_hi_blend = np.logical_and(etas >= 0.99, etas < 1)
+        mask_high = etas >= 1
+        
+        etas = np.atleast_2d(etas).T
+        T = np.zeros_like(etas)
+        
+        # compute thrust
+        T[mask_low] = min_thrs[mask_low] + etas[mask_low]*dTdetas[mask_low]
+        
+        if np.sum(mask_lo_blend) > 0:
```

Comment:
Are the masks like "mask_lo_blend" logical-type or floats? That is, are they full of True/False values or numbers? It looks like its logical-type and I'm not sure if checking it with sum > 0 is robust. Is this meant to be equivalent to checking if any of the entries are True?

Reply:
They're booleans, which are a subtype of integers. This seems to be a fairly standard method for counting the number of True values, but a custom function could be created.

Reply:
They're booleans, which are a subtype of integers. This seems to a typical way of counting the number of True values, but a custom function could also be created.